kolibri/core/assets/src/api_resource.js

```diff
@@ -0,0 +1,68 @@
+const logging = require('loglevel');
+const rest = require('rest');
+const mime = require('rest/interceptor/mime');
+
+const Kolibri = require('kolibri');
+
+const client = rest.wrap(mime);
+
+class Model extends Object {
+  constructor(data, { resourceName, idKey, params } = { idKey: 'id', params: {} }) {
+    super(data);
+    this._resourceName = resourceName;
+    if (typeof this._resourceName !== 'string') {
+      throw new TypeError('resourceName must be defined');
+    }
+    this._idKey = idKey;
+    this._params = params;
+  }
+  fetch() {
+    return new Promise((resolve, reject) => {
+      client({ path: this._url(), params: this._params }).then((response) => {
+        Object.assign(this, response.entity);
```

Comment:
Also, here and more broadly: some very brief inline comments sprinkled around in here could make it easier to parse.